src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt

```diff
@@ -0,0 +1,152 @@
+package com.google.protobuf.gradle
+
+import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.Project
+import org.gradle.api.artifacts.*
+import org.gradle.api.artifacts.dsl.DependencyHandler
+import org.gradle.kotlin.dsl.*
+
+
+fun Project.protobuf(configuration: ProtobufConfigurator.() -> Unit) {
+    project.convention.getPlugin(ProtobufConvention::class.java).protobuf.apply(configuration)
+}
+
+fun ProtobufConfigurator.protoc(closure: ExecutableLocator.() -> Unit) {
+    protoc(delegateClosureOf(closure))
+}
+
+fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
+    plugins(delegateClosureOf<NamedDomainObjectContainer<ExecutableLocator>> {
+        this{ this.closure() }
+    })
+}
+
+fun <T : Any> NamedDomainObjectContainerScope<T>.id(id: String, closure: (T.() -> Unit)? = null) {
+    closure?.let { create(id, delegateClosureOf(it)) }
+        ?: create(id)
+}
+
+fun <T : Any> NamedDomainObjectContainerScope<T>.remove(id: String) {
+    remove(this[id])
+}
+
+fun ProtobufConfigurator.generateProtoTasks(closure: ProtobufConfigurator.GenerateProtoTaskCollection.() -> Unit) {
+    generateProtoTasks(delegateClosureOf(closure))
+}
+
+fun GenerateProtoTask.builtins(closure: NamedDomainObjectContainerScope<GenerateProtoTask.PluginOptions>.() -> Unit) {
+    builtins(delegateClosureOf<NamedDomainObjectContainer<GenerateProtoTask.PluginOptions>> {
+        this{ this.closure() }
+    })
+}
+
+fun GenerateProtoTask.plugins(closure: NamedDomainObjectContainerScope<GenerateProtoTask.PluginOptions>.() -> Unit) {
+    plugins(delegateClosureOf<NamedDomainObjectContainer<GenerateProtoTask.PluginOptions>> {
+        this{ this.closure() }
+    })
+}
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofSourceSet(sourceSet: String): Collection<GenerateProtoTask> =
```

Comment:
Why can't the groovy version of these functions be called directly? Maybe a comment would be helpful.

Reply:
I added a comment to the source but Ill post it here as well.

The method ```generatorProtoTasks``` applies the supplied closure to an
instance of ```GenerateProtoTaskCollection```.

Since ```JavaGenerateProtoTaskCollection``` and ```AndroidGenerateProtoTaskCollection```
each have unique methods, and only one instance in allocated per project, we need a way to statically resolve the available methods. This is a necessity since Kotlin does not have any dynamic method resolution capabilities.

Reply:
Ah, so the problem is the dynamic typing of `GenerateProtoTaskCollection`, since we don't know whether it is `JavaGenerateProtoTaskCollection` or `AndroidGenerateProtoTaskCollection`, kotlin's static typing means we can call neither's methods. 

What do you think about adding some extension methods to `GenerateProtoTaskCollection` like `asJavaTasks` returning the same object casted as `JavaGenerateProtoTaskCollection` and `asAndroidTasks` casting as `AndroidGenerateProtoTaskCollection`. That way, we minimize the amount of extension code that must be kept in sync with the groovy DSL.